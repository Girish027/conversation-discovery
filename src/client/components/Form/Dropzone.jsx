import React, { useMemo, useCallback } from 'react';
import { useDropzone } from 'react-dropzone';
import PropTypes from 'prop-types';

export default function Dropzone(props) {
  const {
    accept,
    saveFile,
    multiple,
    disabled,
    shouldUpload,
    uploadHandler,
    icon: FileIcon,
    acceptedIcon: AcceptIcon,
    acceptStyle,
    activeStyle,
    rejectStyle,
    baseStyle,
  } = props;

  const sizeRestrictions = {};
  if (props.maxSize) {
    sizeRestrictions.maxSize = props.maxSize;
  }
  if (props.minSize) {
    sizeRestrictions.minSize = props.minSize;
  }

  const status = {
    ready: 'ready',
    accepted: 'accepted',
    rejected: 'rejected',
    acceptedFiles: [],
  };

  const state = {
    status: status.ready,
  };

  const eventHandlers = {
    onDrop: useCallback((acceptedFiles) => {
      if (acceptedFiles.length) {
        state.status = status.accepted;
        state.acceptedFiles = acceptedFiles;
        saveFile(acceptedFiles);

        if (shouldUpload) {
          uploadHandler(acceptedFiles);
        }
      }
    }, [saveFile, shouldUpload, state.acceptedFiles, state.status, status.accepted, uploadHandler]),

    onDropRejected: (rejectedFiles) => {
      state.status = status.rejected;
      state.rejectedFiles = rejectedFiles;
    },

  };

  const {
    getRootProps,
    getInputProps,
    isDragActive,
    isDragAccept,
    isDragReject,
    acceptedFiles,
  } = useDropzone({
    ...eventHandlers,
    accept,
    multiple,
    disabled,
    preventDropOnDocument: true,
    ...sizeRestrictions
  });

  const style = useMemo(() => ({
    ...baseStyle,
    ...(isDragActive ? activeStyle : {}),
    ...(isDragAccept ? acceptStyle : {}),
    ...(isDragReject ? rejectStyle : {})
  }), [baseStyle, isDragActive, activeStyle, isDragAccept, acceptStyle, isDragReject, rejectStyle]);

  const renderDetails = (view) => {
    switch (view) {
      case status.ready:
        return `Drag and drop your file
or
BROWSE`;
      case status.accepted: {
        return 'Drop your file here';
      }
      case status.rejected: {
        return `File is not in accepted format.
Please upload ${accept} again.`;
      }
      default: return null;
    }
  };

  return (
    <div {...getRootProps({ style })}>
      <input {...getInputProps()} />
      <div className='dropzone-content'>
        {acceptedFiles.length > 0
          ? acceptedFiles.map((acceptedFile) => (
            <div key={acceptedFile.name}>
              {AcceptIcon
                ? <AcceptIcon height={49} width={49} fill='#004c97' />
                : null
              }
              <div style={{ paddingTop: '10px', whiteSpace: 'pre-wrap' }}>
                {acceptedFile.name}
              </div>
            </div>
          )) : (
            <div>
              {FileIcon
                ? <FileIcon height={49} width={49} />
                : null
              }
              <div style={{ paddingTop: '10px', whiteSpace: 'pre-wrap' }}>
                {isDragAccept && renderDetails(status.accepted)}
                {isDragReject && renderDetails(status.rejected)}
                {!isDragActive && renderDetails(status.ready)}
              </div>
            </div>
          )
        }
      </div>
    </div>
  );
}

const dropzoneStyles = {
  baseStyle: {
    flex: 1,
    display: 'flex',
    flexDirection: 'column',
    height: '90px',
    alignItems: 'center',
    padding: '20px',
    borderWidth: 2,
    borderRadius: 2,
    borderColor: '#eeeeee',
    borderStyle: 'dashed',
    backgroundColor: '#fafafa',
    color: '#bdbdbd',
    outline: 'none',
    transition: 'border .24s ease-in-out'
  },
  activeStyle: {
    borderColor: '#2196f3'
  },

  acceptStyle: {
    borderColor: '#00e676'
  },

  rejectStyle: {
    borderColor: '#ff1744'
  }
};


Dropzone.propTypes = {
  accept: PropTypes.string.isRequired,
  saveFile: PropTypes.func,
  uploadHandler: PropTypes.func,
  icon: PropTypes.func,
  acceptedIcon: PropTypes.func,
  multiple: PropTypes.bool,
  disabled: PropTypes.bool,
  shouldUpload: PropTypes.bool,
  acceptStyle: PropTypes.object,
  activeStyle: PropTypes.object,
  rejectStyle: PropTypes.object,
  baseStyle: PropTypes.object,
  maxSize: PropTypes.number,
  minSize: PropTypes.number,
};

Dropzone.defaultProps = {
  saveFile: () => {},
  uploadHandler: () => {},
  icon: null,
  acceptedIcon: null,
  multiple: false,
  disabled: false,
  shouldUpload: false,
  acceptStyle: dropzoneStyles.acceptStyle,
  activeStyle: dropzoneStyles.activeStyle,
  rejectStyle: dropzoneStyles.rejectStyle,
  baseStyle: dropzoneStyles.baseStyle,
  maxSize: undefined,
  minSize: undefined,
};
