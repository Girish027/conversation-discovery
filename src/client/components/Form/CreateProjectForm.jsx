import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import {
  TextField, Download, Doc, Button
} from '@tfs/ui-components';
import constants from '../../constants';
import Dropzone from './Dropzone';
import CsvIcon from '../Icons/CsvIcon';
import { shouldHandle } from 'utils/KeyboardUtils';
import { downloadFile } from 'utils/FileUtils';

const ERROR_TYPES = {
  DUPLICATE_PROJECT_NAME: 'DUPLICATE_PROJECT_NAME',
  INVALID_FILE_COLUMNS: 'INVALID_FILE_COLUMNS',
};

export default class CreateProjectForm extends PureComponent {
  constructor(props) {
    super(props);
    this.props = props;
    this.onChange = this.onChange.bind(this);

    this.onDownloadKeyPress = this.onDownloadKeyPress.bind(this);
    this.downloadDatasetTemplate = this.downloadDatasetTemplate.bind(this);
  }

  onChange(event) {
    const { value, name } = event.target;
    const { saveChanges } = this.props;
    saveChanges({ [name]: value });
  }

  onDownloadKeyPress(event) {
    if (shouldHandle(event)) {
      this.downloadDatasetTemplate();
    }
    event.preventDefault();
  }

  downloadDatasetTemplate() {
    const { caa } = this.props;
    const locationUrl = constants.serverApiUrls.downloadTemplate(caa);
    downloadFile(locationUrl);
  }

  render() {
    const {
      error, errorType, saveFile, formdata
    } = this.props;

    const { projectName = '', projectDescription = '' } = formdata;
    return (
      <div>
        <div className='formItem'>
          <div className='labelName'>Name</div>
          <TextField
            type='text'
            name='projectName'
            onChange={this.onChange}
            styleOverride={{
              width: '390px',
            }}
            defaultValue={projectName}
          />
        </div>
        <div className='formItem'>
          <div className='labelName'>Description</div>
          <TextField
            type='text'
            name='projectDescription'
            onChange={this.onChange}
            styleOverride={{
              width: '390px',
            }}
            defaultValue={projectDescription}
          />
        </div>
        <div className='formItem'>
          <div className='labelName'>
            <span> Upload Chat Data File </span>
            <hr />
          </div>
          <Button
            onClick={this.downloadDatasetTemplate}
            onKeyDown={this.onDownloadKeyDown}
            type='flat'
            styleOverride={{
              textDecoration: 'underline',
              fontSize: '12px',
              color: '#313f54',
              fontWeight: 'normal'
            }}
          >
            <Download
              width={15}
              height={15}
              fill='#004c97'
              style={{
                verticalAlign: 'middle',
                paddingRight: '10px',
              }}
            />
            Download the template
          </Button>
          <Dropzone
            accept={'text/csv, application/vnd.ms-excel'}
            icon={CsvIcon}
            acceptedIcon={Doc}
            multiple={false}
            saveFile={saveFile}
            maxSize={constants.maxDatasetFileSize}
          />
          { errorType === ERROR_TYPES.INVALID_FILE_COLUMNS && (
            <span className='warning' style={{ height: '15px' }}>
              { error || null }
            </span>
          )}
        </div>
      </div>
    );
  }
}

CreateProjectForm.propTypes = {
  saveFile: PropTypes.func,
  saveChanges: PropTypes.func,
  error: PropTypes.string,
  errorType: PropTypes.string,
  caa: PropTypes.object.isRequired,
  formdata: PropTypes.object,
};

CreateProjectForm.defaultProps = {
  saveFile: () => {},
  saveChanges: () => {},
  error: '',
  errorType: '',
  formdata: {},
};
