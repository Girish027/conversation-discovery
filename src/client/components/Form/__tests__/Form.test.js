import React from 'react';
import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

import Form from '../Form';
import createRunUiSchema from '../../../schema/createRun/uiSchema.json';
import createRunJsonSchema from '../../../schema/createRun/jsonSchema.json';

describe('<Form />', function () {
  let wrapper;


  const props = {
    uiSchema: createRunUiSchema,
    jsonSchema: createRunJsonSchema,
  };

  afterAll(() => {
    jest.clearAllMocks();
  });

  const getShallowWrapper = (propsObj) => shallow(<Form
    {...propsObj}
  />);

  const getMountWrapper = (propsObj) => mount(<Form
    {...propsObj}
  />);

  describe('Creating an instance', () => {
    test('should exist', () => {
      wrapper = getMountWrapper(props);
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    test('renders correctly with default props', () => {
      wrapper = getShallowWrapper(props);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });
});
