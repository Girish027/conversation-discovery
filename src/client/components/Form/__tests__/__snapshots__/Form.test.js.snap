// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> Snapshots renders correctly with default props 1`] = `
<Form
  ErrorList={[Function]}
  ObjectFieldTemplate={[Function]}
  disabled={false}
  fields={
    Object {
      "noop": [Function],
      "text": [Function],
      "textarea": [Function],
      "updown": [Function],
    }
  }
  formContext={Object {}}
  formData={Object {}}
  liveValidate={false}
  noHtml5Validate={false}
  noValidate={false}
  onChange={[Function]}
  onSubmit={[Function]}
  safeRenderCompletion={false}
  schema={
    Object {
      "properties": Object {
        "numOfClusters": Object {
          "default": 2,
          "maximum": 900,
          "minimum": 2,
          "title": "Clusters (2-900):",
        },
        "numOfTurns": Object {
          "default": 1,
          "maximum": 5,
          "minimum": 1,
          "title": "Number of Turns (1-5):",
        },
        "runDescription": Object {
          "maxLength": 150,
          "minLength": 0,
          "title": "Description",
          "type": "string",
        },
        "runName": Object {
          "maxLength": 50,
          "minLength": 4,
          "title": "Name",
          "type": "string",
        },
      },
      "required": Array [
        "runName",
        "numOfClusters",
        "numOfTurns",
      ],
      "title": "",
      "type": "object",
    }
  }
  showErrorList={false}
  transformErrors={[Function]}
  uiSchema={
    Object {
      "numOfClusters": Object {
        "ui:field": "updown",
      },
      "numOfTurns": Object {
        "ui:field": "updown",
      },
      "runDescription": Object {
        "ui:field": "textarea",
      },
      "runName": Object {
        "ui:field": "text",
      },
    }
  }
>
  <div>
    <button
      style={
        Object {
          "display": "none",
        }
      }
      type="submit"
    >
      Submit
    </button>
  </div>
</Form>
`;
